src/python/pants/core/util_rules/gen_rule_test.py

```diff
@@ -0,0 +1,61 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from textwrap import dedent
+
+import pytest
+
+from pants.core.target_types import GenRuleTarget
+from pants.core.util_rules.gen_rule import run_gen_rule
+from pants.core.util_rules.source_files import rules as source_files_rules
+from pants.engine.addresses import Address
+from pants.engine.process import ProcessResult
+from pants.testutil.rule_runner import QueryRule, RuleRunner
+
+
+@pytest.fixture
+def rule_runner() -> RuleRunner:
+    return RuleRunner(
+        rules=[
+            *source_files_rules(),
+            run_gen_rule,
+            QueryRule(ProcessResult, [GenRuleTarget]),
+        ],
+        target_types=[GenRuleTarget],
+    )
+
+
+def assert_gen_rule_result(
+    rule_runner: RuleRunner, address: Address, stdout=None, stderr=None, output_files=None
+) -> None:
+    target = rule_runner.get_target(address)
+    result = rule_runner.request(ProcessResult, [target])
+    if stdout:
+        assert result.stdout == stdout
+    if stderr:
+        assert result.stderr == stderr
+    if output_files:
+        pass
+
+
+def test_simple_message(rule_runner: RuleRunner) -> None:
+    rule_runner.write_files(
+        {
+            "src/BUILD": dedent(
+                """\
+                gen_rule(
+                  name="hello",
+                  sources=["message.txt"],
+                  command="cat message.txt",  # support $(sources) etc similar to what bazel does?
```

Comment:
I'm slightly surprised that `gen_rule` has a `sources` field and no `deps` field... perhaps that's because bazel doesn't have a generic `files` target the way that pants does? But given that we have a generic `files` target, I wonder if this target should consume `dependencies` rather than having sources of its own?

For one thing, assuming that this target is implemented as [codegen](https://www.pantsbuild.org/docs/plugins-codegen) that generates `Files`, having a dependencies field would allow for chaining multiple `gen_rules` together via `dependencies`.

As to your question about a `sources` env var: yea, I think that that makes sense. It allows for usage of globs, for example.

Reply:
The lack of dependencies field is simply by omission to keep the size down. Replacing sources for dependencies makes sense. As you can always depend on a files target to get loose files in.

Reply:
Not entirely sure I get the connection how to support glob usage with a sources env var..?

Reply:
> Not entirely sure I get the connection how to support glob usage with a sources env var..?

If you adjust the example in the PR to use globs, you'll see an issue:
```
sources=["*.txt"],
command="cat *.txt",
```
... won't work for a few reasons:
1. glob expansion usually happens in a shell, and `cat` won't have one here. So it will fail to find a literal file named `*.txt`
2. even if the process being invoked is `bash`, bash's glob expansion will not be equivalent to our own (which tries to match `git`'s syntax).

Having said that! I just realized that there is a bit of a dilemma here. Directly invoking `cat` still wouldn't work with an env var containing a list of sources... because env vars are expanded by shells. All of this might be why Bazel went with requiring that the top level process be `bash`? I could imagine having our own sort of templating/placeholder syntax other than env vars though, maybe...?

Reply:
Mmm, yes, Thatcher could be feasible. I'll expand on the example first, and try to introduce some glob use see how it pans out.

Reply:
I've gone with running the command with `bash -c`. This is the same concept used in Dockerfiles and their `RUN` instruction.